### PR TITLE
Refactor away from PairingEngine + misc type changes

### PR DESCRIPTION
**Pull request by ghost**
Summary of changes:

1. Refactor anything that is only dependent on the scalar field to not depend on PairingEngine (unfortunately this means there's a lot of E::Fr -> F changes in the diff)
2. Move dependence on TEModelParams to other places and not Circuit/StandardComposer. In practice this means (I think) the implementation of the fixed-base ECC gadget is determined later, and the variable-base ECC gadget depends on the params but not the rest of the circuit.
3. Suggested change to remove the TranscriptWrapper type
4. Minor change to Permutation generic types (can be reversed)
5. Some other minor tweaks to types (can be reversed). I was mostly just experimenting to see what might be slightly cleaner.

In summary this PR is not really ready to be merged, but open to input and feedback.

### CODE REVIEW

**Comment by bhgomes** [comment]
Thanks for the PR! For the type parameters, can you use a `where` clause instead of declaring the type constraints inline at the declaration site, it’s easier to read/change this way and keeps it consistent with the rest of the library.

**Comment by bhgomes** [comment]
Also, if you think it's not ready for review you can mark it as draft so we know it's not ready from the PR list.

**Comment by CPerezz** [comment]
Hey @joebebel this looks really promising! Willing to see it completed!

Also, as @bhgomes suggested, if you mark it as Draft for now while the PR is not ready would be nice! :)

**Comment by CPerezz** [comment]
@joebebel 
Since `FFtField` does not imply `PrimeField`, how do we plan to ensure that the field is prime and also extract lots of the associated trait methods info for them?

Also, going this way will probably imply gate-featured verification & prover for IPA and KZG verifications. As now the verification always uses pairings and you don't want to enforce the `PairingEngine`.

**Comment by ghost** [comment]
>  Since `FFtField` does not imply `PrimeField`, how do we plan to ensure that the field is prime and also extract lots of the associated trait methods info for them?

we should just bind to `FftField` when we need those interfaces, and `PrimeField` when we need that. For practical purposes, I don't think any circuit will be over a non prime field. There is some theoretical possibility where non-fft fields might be desirable, but that seems like an extra complication for now. 

> Also, going this way will probably imply gate-featured verification & prover for IPA and KZG verifications. As now the verification always uses pairings and you don't want to enforce the PairingEngine.

As long as we're generic over the field and the `PolynomialCommitment` I think this should work. But it requires some moving of the Rust generics.

**Comment by ghost** [comment]
OK, this is getting closer. There are some problems:

1. For some reason, in `SonicKZG` and `MarlinKZG` the `open` operation on multiple polynomials is not as efficient as the random linear combination before. So more powers of tau in the setup are required. This regression is unfortunate, might require some further attention. I had to increase the setup size to get the tests to pass.
2. Rust type errors when using the `?` operator, as it is unable to convert `ark_poly_commit` errors to `ark_plonk` errors implicitly. I tried several ways, I can't figure out how to get it to work, because the `ark_poly_commit` error is an associated type of `PolynomialCommitment` it's difficult to debug. For now I just `.unwrap()` instead, but this isn't a great long term solution.
3. I added back `TEModelParameters` as a generic for `StandardComposer` and other circuit related structs. It would be ideal to use `ModelParameters` instead (so that short weierstrass inner curves can be used) but after spending some hours trying to figure this out, it's difficult to do in a fully generic way.

I think other than the regression in the efficiency of the polynomial commitment and the error handling, it's ready to review for merge.

**Comment by bhgomes** [comment]
> 2. Rust type errors when using the `?` operator, as it is unable to convert `ark_poly_commit` errors to `ark_plonk` errors implicitly. I tried several ways, I can't figure out how to get it to work, because the `ark_poly_commit` error is an associated type of `PolynomialCommitment` it's difficult to debug. For now I just `.unwrap()` instead, but this isn't a great long term solution.

Whenever an error type is an associated type of a trait, you cannot write a `From` impl since it would actually be generic over every type (since the compiler cannot prove that only some types can inhabit the associated type). So instead, I usually use this pattern:

```rust
let value = fn_returning_ark_poly_commit_error(x, y, z).map_err(ark_plonk::Error::CustomVariant)?;
```

It would be nice if the error could be converted with an `Into<ark_plonk::Error>` impl to remove the `map_err` but as I said above, this is not possible generally (again, this is only because the error we want to map is an associated type, otherwise this does work).

> 3. I added back `TEModelParameters` as a generic for `StandardComposer` and other circuit related structs. It would be ideal to use `ModelParameters` instead (so that short weierstrass inner curves can be used) but after spending some hours trying to figure this out, it's difficult to do in a fully generic way.

I think that the inner curve abstractions in arkworks are lacking here, and in general, there should be a trait that defines the addition and scalar multiplication laws for the curves directly, and the `StandardComposer` just queries that trait for implementing the circuit. Then, when constructing the circuit, you specify which instance of this trait you want, and the circuit is populated with the correct constraints. The arkworks abstractions are too low level to be leveraged to do this, since they are more about specifying the kind of curve (and performing the direct native computation), than specifying the algorithm itself at compile-time.

File: benches/plonk.rs [hunk 1]
```diff
@@ -104,38 +102,64 @@ fn constraint_system_benchmark(c: &mut Criterion) {
...
+                        &pp, key.clone(), &proof, &[], &pi_pos, &label
+=======
                     plonk::circuit::verify_proof(
                         &pp,
                         key.clone(),
                         &proof,
                         &[],
                         &pi_pos,
                         &label,
+>>>>>>> upstream/master
```
Comment (CPerezz) [inline benches/plonk.rs]: There's still merging changes which shouldn't be here.

File: plonk-core/src/constraint_system/logic.rs [hunk 2]
```diff
@@ -293,15 +292,15 @@ where
...
         //     self.variables[&self.w_l[self.n - 1]]
         // );
         // assert_eq!(
         //     self.variables[&b]
-        //         & (E::Fr::from(2u64).pow(&[(num_bits) as u64, 0, 0, 0])
-        //             - E::Fr::one()),
+        //         & (F::from(2u64).pow(&[(num_bits) as u64, 0, 0, 0])
+        //             - F::one()),
         //     self.variables[&self.w_r[self.n - 1]]
         // );
```
Comment (CPerezz) [inline plonk-core/src/constraint_system/logic.rs:305]: We should file an issue for this.
(Not relevant to the PR)

File: plonk-core/src/proof_system/widget/ecc/fixed_base_scalar_mul.rs [hunk 3]
```diff
@@ -86,6 +86,13 @@ where
         checks * separation_challenge
     }
 }
+/*
+impl<F, P> GateConstraint<F> for FixedBaseScalarMul<F, P>
+where
+    F: Field,
+    P: SWModelParameters<BaseField = F>,
+{
+}*/
```
Comment (CPerezz) [inline plonk-core/src/proof_system/widget/ecc/fixed_base_scalar_mul.rs]: Leftover?

File: plonk-core/src/proof_system/widget/mod.rs [hunk 4]
```diff
@@ -141,65 +140,62 @@ where
 #[derive(CanonicalDeserialize, CanonicalSerialize, derivative::Derivative)]
 #[derivative(
     Clone(bound = ""),
-    Debug(bound = ""),
-    Eq(bound = ""),
-    PartialEq(bound = "")
+    //Debug(bound = ""),
+    //Eq(bound = ""),
+    //PartialEq(bound = "")
```
Comment (CPerezz) [inline plonk-core/src/proof_system/widget/mod.rs]: Ditto

File: plonk-core/src/proof_system/widget/mod.rs [hunk 5]
```diff
@@ -466,64 +441,71 @@ mod test {
...
+                q_arith,
+                q_range,
+                q_logic,
+                q_fixed_group_add,
+                q_variable_group_add,
+                left_sigma,
+                right_sigma,
+                out_sigma,
+                fourth_sigma,
+            );
```
Comment (CPerezz) [inline plonk-core/src/proof_system/widget/mod.rs]: Ditto

**Review by CPerezz — COMMENTED** [review]
Just went over the surface and highlighted a few possible leftovers. 

Will go in depth as soon as all is cleared! :)

File: plonk-core/src/proof_system/widget/mod.rs (hunk 4 above)
Comment (ghost) [reply plonk-core/src/proof_system/widget/mod.rs]: `PolynomialCommitment` and `PolynomialCommitment::Commitment` don't implement `Debug, Eq, PartialEq` so we can't derive these traits. Probably another PR for arkworks.

File: plonk-core/src/proof_system/widget/mod.rs (hunk 5 above)
Comment (ghost) [reply plonk-core/src/proof_system/widget/mod.rs]: Also dependent on implementing `Default` in arkworks

File: plonk-core/src/proof_system/widget/ecc/fixed_base_scalar_mul.rs (hunk 3 above)
Comment (ghost) [reply plonk-core/src/proof_system/widget/ecc/fixed_base_scalar_mul.rs]: We need a better way of abstracting the embedded curve type

File: plonk-core/src/circuit.rs [hunk 6]
```diff
@@ -60,82 +46,108 @@ pub trait GeIntoPubInput<T> {
...
 /// positions and the [`VerifierKey`] that the Verifier needs to use.
 #[derive(CanonicalDeserialize, CanonicalSerialize, derivative::Derivative)]
 #[derivative(
     Clone(bound = ""),
-    Debug(bound = ""),
-    Eq(bound = ""),
-    PartialEq(bound = "")
+    //Debug(bound = ""),
+    //Eq(bound = ""),
+    //PartialEq(bound = "")
```
Comment (Pratyush) [inline plonk-core/src/circuit.rs]: You can set the bound to `F: PrimeField, VerifierKey<F, PC>: foo`, where `foo \in {Debug, Eq, PartialEq}`.

**Comment by Pratyush** [comment]
> For some reason, in SonicKZG and MarlinKZG the open operation on multiple polynomials is not as efficient as the random linear combination before. So more powers of tau in the setup are required. This regression is unfortunate, might require some further attention. I had to increase the setup size to get the tests to pass.

This should not be the case unless you're using hiding and/or strict degree bounds (which plonk doesn't need, AFAIK). I can hop on a call to help debug. btw you can turn on the `print-trace` feature on `ark-poly-commit`; that should give you a sense of why opening requires a larger SRS.

File: plonk-core/src/circuit.rs (hunk 6 above)
Comment (ghost) [reply plonk-core/src/circuit.rs]: Hmm, I don't quite understand. The problem is that `PolynomialCommitment` and `PolynomialCommitment::Commitment` don't impl `Copy, Debug, Eq, PartialEq`. So no matter how we put the bounds in it'll never be satisfied.

File: benches/plonk.rs (hunk 1 above)
Comment (ghost) [reply benches/plonk.rs]: Hah, oops.

**Comment by ghost** [comment]
> This should not be the case unless you're using hiding and/or strict degree bounds (which plonk doesn't need, AFAIK). I can hop on a call to help debug. btw you can turn on the `print-trace` feature on `ark-poly-commit`; that should give you a sense of why opening requires a larger SRS.

I don't know what I changed to fix it, but now it functions correctly. It wasn't a problem with the aggregation, though; the error was `TooManyCoefficients { num_coefficients: 80, num_powers: 23 }`. In this case the largest degree of the aggregated polynomials, `t_4_poly` was correctly 80, but only a degree 23 setup was generated. Strangely, `t_4_poly` was degree 23 in the previous proof in the same test. So somehow the CRS was not correctly sized.

**Comment by LukePearson1** [comment]
When this ready for a review, even if still in draft, could you let us know and I'll give a first round? 😄  @joebebel

File: plonk-core/src/proof_system/widget/mod.rs (hunk 4 above)
Comment (ghost) [reply plonk-core/src/proof_system/widget/mod.rs]: Fixed, I just added the correct `bound=`

**Comment by ghost** [comment]
I believe this branch is now non-inferior (passes all tests, etc) to the main branch, so it is "ready for review".

There is one main deficiency, which is that it performs *two* `PC::check` operations (whereas the original effectively just did one). So slightly worse performance.

Thanks @Pratyush, I figured out what you meant about the trait bounding and it seems to work well. Also thanks @bhgomes, the error handling is better now.

File: plonk-core/src/proof_system/widget/mod.rs (hunk 4 above)
Comment (CPerezz) [reply plonk-core/src/proof_system/widget/mod.rs]: Thanks!!

File: plonk-core/src/constraint_system/boolean.rs [hunk 7]
```diff
@@ -7,13 +7,13 @@
...
-use num_traits::{One, Zero};
+use ark_ec::ModelParameters;
+use ark_ff::{FftField, PrimeField};
 
-impl<E, P> StandardComposer<E, P>
+impl<F, P> StandardComposer<F, P>
 where
-    E: PairingEngine,
-    P: TEModelParameters<BaseField = E::Fr>,
+    F: FftField + PrimeField,
```
Comment (davidnevadoc) [inline plonk-core/src/constraint_system/boolean.rs]: `PrimeField` already requires `FftField` so we can just leave it as `PrimeField`.
The same goes for the rest of appearances in the PR but just commented once.

File: plonk-core/src/circuit.rs [hunk 8]
```diff
@@ -605,12 +595,29 @@ mod test {
...
     #[allow(non_snake_case)]
     fn test_full_on_Bls12_381() -> Result<(), Error> {
-        test_full::<Bls12_381, ark_ed_on_bls12_381::EdwardsParameters>()
+        test_full::<
+            <Bls12_381 as PairingEngine>::Fr,
+            ark_ed_on_bls12_381::EdwardsParameters,
+            crate::commitment::KZG10<Bls12_381>,
+        >()?;
+
+        test_full::<
```
Comment (davidnevadoc) [inline plonk-core/src/circuit.rs:611]: I think it's best to separate in 2 tests :
`test_full_on_Bls12_381` and `test_full_on_Bls12_381_ipa`

File: plonk-core/src/constraint_system/arithmetic.rs [hunk 9]
```diff
@@ -460,8 +456,35 @@ mod test {
...
+            test_incorrect_big_arith_gate
+        ],
+        [] => (
+            Bls12_381, ark_ed_on_bls12_381::EdwardsParameters
         )
     );
+
+    #[test]
+    #[allow(non_snake_case)]
+    fn test_ipa() {
```
Comment (davidnevadoc) [inline plonk-core/src/constraint_system/arithmetic.rs]: Is this test necessary?

**Review by davidnevadoc — APPROVED** [review]
What an amazing PR! 

Just highlighted a couple of nits but overall I love the improvements introduced. Great job! :)

File: plonk-core/src/error.rs [hunk 10]
```diff
@@ -78,7 +78,23 @@ pub enum Error {
...
+pub fn to_pc_error<F, PC>(error: PC::Error) -> Error
+where
+    F: ark_ff::Field,
+    PC: ark_poly_commit::PolynomialCommitment<
+        F,
+        ark_poly::univariate::DensePolynomial<F>,
+    >,
+{
+    Error::PCError {
+        error: format!("Polynomial Commitment Error: {:?}", error),
```
Comment (CPerezz) [inline plonk-core/src/error.rs:97]: Why do we need this function? We have the `From` impl already.

Not sure if is a leftover or has a purpose that I don't see now.

File: plonk-core/src/proof_system/preprocess.rs [hunk 11]
```diff
@@ -214,14 +224,18 @@ where
...
+        commit_key: &PC::CommitterKey,
+        transcript: &mut Transcript,
+        _pc: PhantomData<PC>,
+    ) -> Result<widget::VerifierKey<F, PC>, Error>
+    where
+        PC: HomomorphicCommitment<F>,
+    {
         let (verifier_key, _, _) =
-            self.preprocess_shared(commit_key, transcript)?;
+            self.preprocess_shared(commit_key, transcript, _pc).unwrap();
```
Comment (CPerezz) [inline plonk-core/src/proof_system/preprocess.rs]: Why do we unwrap and then wrap the result into a `Result` again?

Why not just doing:
```suggestion
            self.preprocess_shared(commit_key, transcript, _pc)
```

File: plonk-core/src/proof_system/preprocess.rs [hunk 12]
```diff
@@ -295,115 +313,122 @@ where
...
-
-        let fourth_sigma_poly_commit =
-            KZG10::<E, DensePolynomial<E::Fr>>::commit(
-                commit_key,
-                &fourth_sigma_poly,
-                None,
-                None,
-            )?;
+        )
+        .map_err(to_pc_error::<F, PC>)?;
```
Comment (CPerezz) [inline plonk-core/src/proof_system/preprocess.rs:349]: Doesn't the `From` impl do the same without needing to call this fn?

File: plonk-core/src/proof_system/proof.rs [hunk 13]
```diff
@@ -46,73 +43,80 @@ use rand_core::OsRng;
...
+
+    /// Batch opening of the commitments
+    pub aw_opening: PC::Proof,
+
+    /// Batch openings of the shifted commitments
+    pub saw_opening: PC::Proof,
 
     /// Commitment to the opening proof polynomial.
-    pub(crate) w_z_comm: Commitment<E>,
+    //pub(crate) w_z_comm: PC::Commitment,
```
Comment (CPerezz) [inline plonk-core/src/proof_system/proof.rs]: leftover?

File: plonk-core/src/proof_system/proof.rs [hunk 14]
```diff
@@ -46,73 +43,80 @@ use rand_core::OsRng;
...
 
     /// Commitment to the shifted opening proof polynomial.
-    pub(crate) w_zw_comm: Commitment<E>,
+    //pub(crate) w_zw_comm: PC::Commitment,
```
Comment (CPerezz) [inline plonk-core/src/proof_system/proof.rs]: ditto

File: plonk-core/src/proof_system/proof.rs [hunk 15]
```diff
@@ -46,73 +43,80 @@ use rand_core::OsRng;
...
+        verifier_key: &PC::VerifierKey,
+        pub_inputs: &[F],
+    ) -> Result<(), Error>
+    where
+        P: TEModelParameters<BaseField = F>,
+    {
         let domain =
-            GeneralEvaluationDomain::<E::Fr>::new(plonk_verifier_key.n)
-                .unwrap();
+            GeneralEvaluationDomain::<F>::new(plonk_verifier_key.n).unwrap();
```
Comment (CPerezz) [inline plonk-core/src/proof_system/proof.rs]: You should not need to unwrap now that you have the `From` impl

File: plonk-core/src/proof_system/proof.rs [hunk 16]
```diff
@@ -46,73 +43,80 @@ use rand_core::OsRng;
...
 
     /// Commitment to the quotient polynomial.
-    pub(crate) t_4_comm: Commitment<E>,
+    pub(crate) t_4_comm: PC::Commitment,
+
+    /// Batch opening of the commitments
+    pub aw_opening: PC::Proof,
+
+    /// Batch openings of the shifted commitments
+    pub saw_opening: PC::Proof,
```
Comment (CPerezz) [inline plonk-core/src/proof_system/proof.rs]: Can we add in the comments explicitly that is the `aggregated_witness` opening?

I think instead of `commitments` you meant `witnesses`

File: plonk-core/src/proof_system/prover.rs [hunk 17]
```diff
@@ -278,24 +268,29 @@ where
...
+            label_polynomial!(w_4_poly),
+        ];
 
         // Commit to witness polynomials.
-        let w_l_poly_commit = KZG10::commit(commit_key, &w_l_poly, None, None)?;
-        let w_r_poly_commit = KZG10::commit(commit_key, &w_r_poly, None, None)?;
-        let w_o_poly_commit = KZG10::commit(commit_key, &w_o_poly, None, None)?;
-        let w_4_poly_commit = KZG10::commit(commit_key, &w_4_poly, None, None)?;
+        let (w_commits, w_rands) = PC::commit(commit_key, w_polys.iter(), None)
+            .map_err(to_pc_error::<F, PC>)?;
```
Comment (CPerezz) [inline plonk-core/src/proof_system/prover.rs:282]: shouldn't `.map_err(|e| Error::from(e))?;` do the work?

File: plonk-core/src/test.rs [hunk 18]
```diff
@@ -18,15 +88,38 @@ macro_rules! batch_test {
...
+                fn [< $test_set _on_ $curve _ipa>]() {
+                    $test_set::<<$curve as ark_ec::PairingEngine>::Fr, $params, ark_poly_commit::ipa_pc::InnerProductArgPC<<$curve as ark_ec::PairingEngine>::G1Affine, blake2::Blake2s, ark_poly::univariate::DensePolynomial<<$curve as ark_ec::PairingEngine>::Fr>>>()
+                }
+            )*
+            $(
+                #[test]
+                #[should_panic]
+                #[allow(non_snake_case)]
+                fn [< $test_panic_set _on_ $curve _ipa>]() {
+                    $test_panic_set::<<$curve as ark_ec::PairingEngine>::Fr, $params, ark_poly_commit::ipa_pc::InnerProductArgPC<<$curve as ark_ec::PairingEngine>::G1Affine, blake2::Blake2s, ark_poly::univariate::DensePolynomial<<$curve as ark_ec::PairingEngine>::Fr>>>()
```
Comment (CPerezz) [inline plonk-core/src/test.rs]: Considering that we test with IPA the same exact number of test cases than with KZG, wouldn't be easier to include the call to `batch_test_ipa!` inside the `batch_test!` macro? 

We would reduce a lot of code duplicity.

File: plonk-core/src/transcript.rs [hunk 19]
```diff
@@ -7,91 +7,76 @@
...
 /// For convenience
-pub(crate) trait TranscriptProtocol<E>
-where
-    E: PairingEngine,
-{
-    /// Append a `commitment` with the given `label`.
-    fn append_commitment(&mut self, label: &'static [u8], comm: &Commitment<E>);
+pub(crate) trait TranscriptProtocol {
+    /// Append an `item` with the given `label`.
+    fn append(&mut self, label: &'static [u8], item: &impl CanonicalSerialize);
```
Comment (CPerezz) [inline plonk-core/src/transcript.rs:21]: That's a really clever idea! I love it!

File: plonk-core/src/proof_system/preprocess.rs [hunk 20]
```diff
@@ -295,115 +313,122 @@ where
...
+                    out_sigma_poly.clone(),
+                    None,
+                    None,
+                ),
+                LabeledPolynomial::new(
+                    "fourth_sigma_poly".to_owned(),
+                    fourth_sigma_poly.clone(),
+                    None,
+                    None,
+                ),
```
Comment (CPerezz) [inline plonk-core/src/proof_system/preprocess.rs]: Use the  `label_polynomial!` macro instead.

**Review by CPerezz — CHANGES_REQUESTED** [review]
Some nits and questions.

Aside from that, this is a brilliant work @joebebel. 
Congratulations and big thanks for this huge improvement! :1st_place_medal:

File: plonk-core/src/constraint_system/boolean.rs (hunk 7 above)
Comment (ghost) [reply plonk-core/src/constraint_system/boolean.rs]: Huh, I never realized this (maybe because I assumed that not every `PrimeField` is FFT friendly)

File: plonk-core/src/constraint_system/arithmetic.rs (hunk 9 above)
Comment (ghost) [reply plonk-core/src/constraint_system/arithmetic.rs]: leftover from when I was just testing whether swapping IPA for KZG10 worked at all, so I'll remove it

File: plonk-core/src/proof_system/preprocess.rs (hunk 11 above)
Comment (ghost) [reply plonk-core/src/proof_system/preprocess.rs]: ah, leftover from when `?` didn't work temporarily

File: plonk-core/src/proof_system/proof.rs (hunk 15 above)
Comment (ghost) [reply plonk-core/src/proof_system/proof.rs]: We should do something like:
```
.ok_or(Error::InvalidEvalDomainSize {
                log_size_of_group: plonk_verifier_key.n.trailing_zeros(),
                adacity: <F as FftField>::FftParams::TWO_ADACITY,
            })?
```
instead of `unwrap()` but I'm having some issues implementing this.
 I could just remove the parameters from `InvalidEvalDomainSize` or something quick, or leave it as `unwrap` and open another issue/PR for this case (since it's rather independent)

File: plonk-core/src/proof_system/prover.rs (hunk 17 above)
Comment (ghost) [reply plonk-core/src/proof_system/prover.rs:282]: It doesn't like impl `From` for associated types, I guess because it has no idea if there might exist another `From` impl for the actual type. I'm open for suggestions but I don't have a clean proposal other than what is implemented here.

File: plonk-core/src/test.rs (hunk 18 above)
Comment (ghost) [reply plonk-core/src/test.rs]: merged into `batch_test` then.

File: plonk-core/src/error.rs (hunk 10 above)
Comment (ghost) [reply plonk-core/src/error.rs:97]: Unfortunately the `From` trait doesn't seem to like associated types. So we can `from` `ark_poly_commit::Error` but when I try to `From<PC::Error>`, maybe it doesn't know what the associated type is, or if there is already a `From` impl, etc.
Open to suggestions, I'm not quite sure how to handle this case.

File: plonk-core/src/proof_system/preprocess.rs (hunk 12 above)
Comment (ghost) [reply plonk-core/src/proof_system/preprocess.rs:349]: No, unfortunately again, this is throwing off an error of type `PC::Error` and not `ark_poly_commit::Error` so the `From` doesn't handle it.

File: plonk-core/src/proof_system/proof.rs (hunk 16 above)
Comment (ghost) [reply plonk-core/src/proof_system/proof.rs]: Ok, I think I changed it to what you suggested?

File: plonk-core/src/transcript.rs (hunk 19 above)
Comment (ghost) [reply plonk-core/src/transcript.rs:21]: Yeah this is one of the cases where working with the type system makes it much nicer/easier.

File: plonk-core/src/proof_system/proof.rs (hunk 15 above)
Comment (CPerezz) [reply plonk-core/src/proof_system/proof.rs]: Sure. Don't want to extend this PR opened more. I've added an issue for it so that we can discuss better the error handling here.

File: plonk-core/src/proof_system/proof.rs (hunk 16 above)
Comment (CPerezz) [reply plonk-core/src/proof_system/proof.rs]: Yep!!! Thanks!!

File: plonk-core/src/proof_system/prover.rs (hunk 17 above)
Comment (CPerezz) [reply plonk-core/src/proof_system/prover.rs:282]: Will give it a thought. Filed #86

File: plonk-core/src/proof_system/preprocess.rs (hunk 12 above)
Comment (CPerezz) [reply plonk-core/src/proof_system/preprocess.rs:349]: Filed #86

File: plonk-core/src/error.rs (hunk 10 above)
Comment (CPerezz) [reply plonk-core/src/error.rs:97]: Filed #86 to give it a thought without blocking this! :)

File: plonk-core/src/test.rs (hunk 18 above)
Comment (CPerezz) [reply plonk-core/src/test.rs]: Thanks!!!

**Review by CPerezz — APPROVED** [review]
A couple unused things in `proof.rs` 
Once fixed the CI should pass. 

Once it passes, feel free to merge! 
Awesome job with this PR @joebebel !!!!!

**Comment by ghost** [comment]
OK, ready to merge
